docs/python-grader/index.md

```diff
@@ -0,0 +1,214 @@
+ Checks that a numpy array has the same shape and datatype as the reference solution, also checks values to see if they are close using specified `rtol` and `atol`.
+- `check_list(name, ref, data, entry_type=None)`
+ Checks that a list has the same length as the reference solution.  If `entry_type` is not `None`, can optionally check if each element has that type.  Does *not* check values against the reference.
+- `check_tuple(name, ref, data)`
+ Checks that a tuple has the same length and values as the reference solution.
+- `check_scalar(name, ref, data, rtol=1e-5, atol=1e-8)`
+ Checks that a scalar value is close to the reference solution using specified `rtol` and `atol`.
+- `call_user(f, *args, **kwargs)`
+ Calls a user defined function with specific `args` and `kwargs`.
+- `check_dataframe(name, ref, data, subset_columns=None)`
+ Checks that a Pandas DataFrame has the same contents as the reference solution.  Can optionally check for a subset of columns by giving a list of column names to `subset_columns`.
+
+Most of these functions have a `accuracy_critical` and `report_failure` keyword argument.  If `accuracy_critical` is set to true (default), the grading job will halt if this check fails (similar to an assert).  If `report_failure` is true (default), feedback will be displayed to the student if this check fails.  The `name` argument is used when displaying feedback.
+
+## General Tips and Gotchas
+
+Note that the first argument of the `feedback.check_xx` functions is the name of the variable being checked, this will show up in the grader feedback if the student answers this problem incorrectly.
+
+Be careful not to switch the ordering of the student and reference arguments.  The student answer is subject to more strict type checking, and there have been instances in the past where the grader has been broken by poorly formatted student answers.
+
+## Banning/Disallowing library functions
+
+One can hook into library in the setup code to disallow students from accessing certain functions.  This example is taken from the `demoAutograderNumpy` question.
+
+By creating a `not_allowed` function that only raises an error when called:
+
+```python
+def not_allowed(*args, **kwargs):
+    raise RuntimeError("Calling this function is not allowed")
+```
+
+and then setting library functions equal to `not_allowed`:
+
+```python
+numpy.linalg.inv = not_allowed
+numpy.linalg.pinv = not_allowed
+```
+
+the `inv` and `pinv` functions will be effectively banned from use.  This will survive library reimports.
+
+# Overview
+
+![](grader-structure.png)
+
+# Security
+
+On the start of a grading job, the script `run.sh` is run with `root` privileges.  This will initialise directories and copy grading files into the correct location, delete itself, and generate a secret filename for the grading results using `uuidgen` which is saved into `output-fname.txt`. Then the actual grading script will then be run (`pltest.py`) as the user `ag`.
```

Comment:
```suggestion
At the start of a grading job, the script `run.sh` is run with `root` privileges.  This will initialise directories and copy grading files into the correct location, delete itself, and generate a secret filename for the grading results using `uuidgen` which is saved into `output-fname.txt`. Then the actual grading script will then be run (`pltest.py`) as the user `ag`.
```